include/dlaf/eigensolver/bt_band_to_tridiag/impl.h

```diff
@@ -562,4 +639,291 @@ void BackTransformationT2B<B, D, T>::call(const SizeType band_size, Matrix<T, D>
+  const matrix::Distribution dist_ws_w2({nlocal_ws * b, mat_e.size().cols()},
+                                        {b, mat_e.blockSize().cols()});
+
+  constexpr std::size_t n_workspaces = 2;
+
+  RoundRobin<Panel<Coord::Col, T, D>> t_panels(n_workspaces, dist_ws_hh);
+  RoundRobin<Panel<Coord::Col, T, D>> hh_panels(n_workspaces, dist_ws_hh);
+
+  RoundRobin<Panel<Coord::Col, T, D>> v_panels(n_workspaces, dist_ws_v);
+  RoundRobin<Panel<Coord::Col, T, D>> w_panels(n_workspaces, dist_ws_v);
+
+  RoundRobin<Panel<Coord::Row, T, D>> w2_panels(n_workspaces, dist_ws_w2);
+  RoundRobin<Panel<Coord::Row, T, D>> w2tmp_panels(n_workspaces, dist_ws_w2);
+
+  HHManager<B, D, T> helperBackend(b, n_workspaces, dist_ws_hh, dist_ws_v);
+
+  // Note: This distributed algorithm encompass two communication categories:
+  // 1. exchange of HH: broadcast + send p2p
+  // 2. reduction for computing W2: all reduce p2p
+  // P2P communication can happen out of order since they can be matched via tags, but this is not
+  // possible for collective operations such as the broadcast.
+  //
+  // For this reason, communications of the phase 1 will be ordered with a pipeline. Instead, for the
+  // second part, with the aim to not over constrain execution of the update, no order will be
+  // enforced by relying solely on tags.
+  common::Pipeline<comm::Communicator> mpi_chain_row(grid.rowCommunicator().clone());
+  common::Pipeline<comm::Communicator> mpi_chain_col(grid.colCommunicator().clone());
+  const auto mpi_col_comm = ex::just(grid.colCommunicator().clone());
+
+  const SizeType idx_last_sweep_b = (nrSweeps<T>(mat_hh.size().cols()) - 1) / b;
+  const SizeType maxsteps_b = nrStepsForSweep(0, mat_hh.size().rows(), b);
+
+  // Note: Next two nested `for`s describe a special order loop over the matrix, which allow to
+  // better schedule communications considering the structure of the algorithm.
+  //
+  // Each element depends on:
+  // - top
+  // - bottom-right
+  // - right
+  //
+  // This basic rule for dependencies can be described collectively as a mechanism where elements are
+  // "unlocked" in different epochs, which forms a pattern like if the matrix get scanned not
+  // perpendicularly to their main axis, but instead it gets scanned by a slightly skewed line that goes
+  // from top right to bottom left.
+  //
+  //  5 x x x x
+  //  6 4 x x x
+  //  7 5 3 x x
+  //  8 6 4 2 x
+  //  9 7 5 3 1
+  //
+  // Elements of the same epoch are somehow "independent" and so they can potentially run in parallel,
+  // given that previous epoch has been completed. Since scheduling happens sequentially, elements
+  // of the same epoch will be ordered starting from top-most one, resulting in
+  //
+  //  7  x x x x
+  // 10  5 x x x
+  // 12  8 3 x x
+  // 14 11 6 2 x
+  // 15 13 9 4 1
+  //
+  // TODO test invert order intra-epoch if it gives a performance hit/improvement
```

Comment:
In general we don't expect this makes a big impact...I would go for opening an issue about this, so that we can take this into account when we will be in the optimisation phase.

Reply:
Ok, indeed, an issue sounds good.